Drop Python 3.7, 3.8, 3.9 support

ivanovmg opened the pull request:
This PR drops support for Python 3.7, 3.8 and 3.9 versions.

Here we updated python version in `pyproject.toml` and then updated `poetry.lock` file as described below (please correct me if it is totally wrong way of doing things).

Update to `poetry.lock` file:
1. Bump python version to 3.10+ (previous commit)
2. Re-build poetry.lock file as shown below:
```
$ rm poetry.lock
$ poetry install
```

## Checklist

<!-- Please check everything that applies: -->

- [ ] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [ ] I have created at least one test case for the changes I have made
- [ ] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Format is:
- Closes #496

🙏 Please, if you or your company finds `dry-python` valuable, help us sustain the project by sponsoring it transparently on https://github.com/sponsors/dry-python. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

ivanovmg committed:
Fix poetry install script

Closes #494

References:
- https://python-poetry.org/docs/#installing-with-the-official-installer

ivanovmg committed:
Bump minimum python version to 3.10

ivanovmg committed:
Update poetry.lock file

Repro steps:
1. Bump python version to 3.10+ (previous commit)
2. Re-build poetry.lock file as shown below:

```
$ rm poetry.lock
$ poetry install
```

ivanovmg committed:
Add TODO related to dropping the entire test module

ivanovmg committed:
Update CHANGELOG

ivanovmg committed:
Update python versions in github CI

1. Dropped 3.7-3.9
2. Added 3.11-3.13

sobolevn commented on tests/test_typeclass/test_typed_dict.py:
```diff
@@ -8,6 +8,7 @@
 if sys.version_info[:2] >= (3, 9):  # noqa: C901
```
You can remove that now, because there are no python versions that won't execute this branch.

sobolevn reviewed (commented):
Please, take a look at other `sys.version_info` checks

sobolevn commented:
you can use `pip-check` lib from PyPI to find outdate packages and update them all at once in `pyproject.toml`

ivanovmg committed:
Drop test_typed_dict.py

Previously in this test module the actual tests would be
executed only for python 3.9 and earlier.
Since we drop support for these versions, then these tests
are not necessary anymore.

ivanovmg committed:
Bump dev requirements

sobolevn commented:
I pushed a fix for this problem to `master`, please, rebase

ivanovmg commented:
> I pushed a fix for this problem to `master`, please, rebase

OK, thanks! I tried to fix it myself, but now I will rebase

ivanovmg committed:
Avoid inheritance from object

Avoid inheritance from object
to fix WPS306 linting errors.

```
WPS306 Found explicit `object` base class: ...
```

ivanovmg committed:
Re-implement union_items function

Since mypy v0.980 this function has been removed from `mypy.types`.
This this commit we re-implement the same version
present in v0.970.

ivanovmg committed:
Fix missing type parameters for Callable

ivanovmg committed:
Fix import of Protocol

ivanovmg committed:
Add docstrings to satisfy D101

ivanovmg commented:
It turns out that there are multiple mypy check failures with the newest mypy version.
I tried to fix some of them, but it goes beyond my experience and knowledge of type checking.

@sobolevn please suggest the way forward.
I suppose that some of them can be silenced, considering a dynamic nature of the `typeclass` implementation, but most of them will still have to be handled somehow.

sobolevn commented:
You can use older mypy for now, I will fix these issues later :)

ivanovmg committed:
Downgrade mypy to v0.942

ivanovmg committed:
Revert "Fix missing type parameters for Callable"

This reverts commit e6d74cd4a1e3c67c84080a833866fe45050bed44.

ivanovmg commented:
> You can use older mypy for now, I will fix these issues later :)

I reverted back to mypy v0.942, but surprisingly the errors persist (even more errors appeared).

**Update**

I noticed that when I drop `strict=true` in `setup.cfg`, then `mypy` checks are successful, but then flake8 complains.
![image](https://github.com/user-attachments/assets/07526edf-c66a-48b9-a2ad-257a8cde64d7)

With `strict = true` we have multiple `mypy` failures.

ivanovmg committed:
Turn off mypy strict rule

sobolevn reviewed (commented):
You can delete `nitpick` section and dependency from `pyproject.toml`

sobolevn committed:
Update pyproject.toml
